File: dataforge-plots/plots-fx/src/main/java/hep/dataforge/plots/fx/PlotContainer.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package hep.dataforge.plots.fx;

import hep.dataforge.description.DescriptorUtils;
import hep.dataforge.description.NodeDescriptor;
import hep.dataforge.fx.FXObject;
import hep.dataforge.fx.FXUtils;
import hep.dataforge.fx.configuration.ConfigEditor;
import hep.dataforge.meta.ConfigChangeListener;
import hep.dataforge.meta.Configuration;
import hep.dataforge.meta.Meta;
import hep.dataforge.plots.Plot;
import hep.dataforge.plots.Plottable;
import hep.dataforge.values.Value;
import javafx.application.Platform;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.DoubleProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleDoubleProperty;
import javafx.beans.value.ObservableValue;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.scene.Node;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;
import javafx.stage.Stage;
import javafx.stage.WindowEvent;
import org.slf4j.LoggerFactory;

import java.net.URL;
import java.util.*;

/**
 * FXML Controller class
 *
 * @author Alexander Nozik
 */
public class PlotContainer implements Initializable, FXObject {

    private AnchorPane root;
    @FXML
    private BorderPane plotPane;
    @FXML
    private VBox sideBar;
    @FXML
    private ListView<Plot> plottableslList;
    @FXML
    private Button optionsPannelButton;
    @FXML
    private Button frameOptionsButton;
    @FXML
    private SplitPane split;
    @FXML
    private ProgressIndicator progressIndicator;

    private FXPlotFrame plot;
    private Map<Configuration, Stage> configWindows = new HashMap<>();
    private BooleanProperty sidebarVisibleProperty = new SimpleBooleanProperty(true);
    private double lastDividerPosition = -1;
    private DoubleProperty progressProperty = new SimpleDoubleProperty(1.0);

    public PlotContainer() {
        FXMLLoader loader = new FXMLLoader(getClass().getResource("/fxml/PlotContainer.fxml"));
        root = new AnchorPane();
        loader.setRoot(root);
        loader.setController(this);

        try {
            loader.load();
        } catch (Exception ex) {
            LoggerFactory.getLogger("FX").error("Error during fxml initialization", ex);
            throw new Error(ex);
        }

        progressIndicator.progressProperty().bind(progressProperty);
        progressIndicator.visibleProperty().bind(progressProperty.lessThan(1.0));
    }

    public static PlotContainer anchorTo(AnchorPane pane) {
        PlotContainer container = new PlotContainer();
        pane.getChildren().add(container.getPane());
        AnchorPane.setBottomAnchor(container.getPane(), 0d);
        AnchorPane.setTopAnchor(container.getPane(), 0d);
        AnchorPane.setLeftAnchor(container.getPane(), 0d);
        AnchorPane.setRightAnchor(container.getPane(), 0d);
        return container;
    }

    public static PlotContainer centerIn(BorderPane pane) {
        PlotContainer container = new PlotContainer();
        pane.setCenter(container.getPane());
        return container;
    }

    /**
     * The root node of container
     *
     * @return
     */
    public Parent getPane() {
        return root;
    }

    @Override
    public Node getFXNode() {
        return getPane();
    }

    /**
     * Initializes the controller class.
     */
    @Override
    public void initialize(URL url, ResourceBundle rb) {
        plottableslList.setCellFactory((ListView<Plot> param) -> new PlottableListCell());
        plottableslList.getSelectionModel().setSelectionMode(SelectionMode.MULTIPLE);
        root.setMaxSize(Double.MAX_VALUE, Double.MAX_VALUE);
        split.getDividers().get(0).positionProperty().addListener(
                (ObservableValue<? extends Number> observable, Number oldValue, Number newValue) -> {
                    sidebarVisibleProperty.set(newValue.doubleValue() < 0.99);
                    if (newValue.doubleValue() < 0.98) {
                        lastDividerPosition = newValue.doubleValue();
                    }
                });

        sidebarVisibleProperty.addListener(
                (ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) -> {
                    if (newValue) {
                        optionsPannelButton.setText(">>");
                    } else {
                        optionsPannelButton.setText("<<");
                    }
                });

        optionsPannelButton.setOnAction((ActionEvent event) -> {
            if (sidebarVisibleProperty.get()) {
                split.setDividerPosition(0, 1d);
            } else if (lastDividerPosition > 0) {
                split.setDividerPosition(0, lastDividerPosition);
            } else {
                split.setDividerPosition(0, 0.8);
            }
        });
        split.setDividerPositions(1.0);
    }

    public void setSideBarExpanded(boolean expanded) {
        this.sidebarVisibleProperty.set(expanded);
    }

    public void setSideBarPosition(double position) {
        split.setDividerPositions(position);
    }

    /**
     * A list of getPlottables in the sidebar
     *
     * @return
     */
    public ListView<Plot> getPlottableslListView() {
        return plottableslList;
    }

    /**
     * An sideBar VBox
     *
     * @return
     */
    protected VBox getSideBar() {
        return sideBar;
    }

    /**
     * Add nodes to the end of the sideBar
     *
     * @param nodes
     */
    public void addToSideBar(Node... nodes) {
        getSideBar().getChildren().addAll(nodes);
    }

    /**
     * Insert nodes in sideBar at specific index
     *
     * @param index
     * @param nodes
     */
    public void addToSideBar(int index, Node... nodes) {
        getSideBar().getChildren().addAll(index, Arrays.asList(nodes));
    }

    public FXPlotFrame getPlot() {
        return plot;
    }

    /**
     * Set plot to display in this container
     *
     * @param plot
     */
    public void setPlot(FXPlotFrame plot) {
        this.plot = plot;

        ObservableList<Plot> plots = FXCollections.observableArrayList();

        plot.getPlots().forEach(it-> {
            if(it instanceof Plot){
                plots.add((Plot) it);
            }
        });

        FXUtils.runNow(() -> {
            plotPane.getChildren().retainAll(optionsPannelButton);
            plotPane.setCenter(plot.getFXNode());

            plottableslList.setItems(plots);
        });
    }

    /**
     * remove plot from container
     */
    public void removePlot() {
        this.plot = null;
        FXUtils.runNow(() -> {
            plotPane.getChildren().retainAll(optionsPannelButton);
            plottableslList.setItems(FXCollections.emptyObservableList());
        });
    }

    protected void setupSideBar() {
        FXUtils.runNow(() -> plottableslList.getItems().clear());
    }

    /**
     * Display configurator in separate scene
     *
     * @param config
     * @param desc
     */
    private void displayConfigurator(String header, Configuration config, NodeDescriptor desc) {
        Stage stage;
        if (configWindows.containsKey(config)) {
            stage = configWindows.get(config);
        } else {
            stage = new Stage();
            Scene scene = new Scene(ConfigEditor.build(config, desc));
            stage.setScene(scene);
            stage.setHeight(400);
            stage.setWidth(400);
            stage.setTitle(header);
            stage.setOnCloseRequest((WindowEvent event) -> configWindows.remove(config));
            stage.initOwner(root.getScene().getWindow());
            configWindows.put(config, stage);
        }
        stage.show();
        stage.toFront();
    }

    @FXML
    private void onFrameOptionsClick(ActionEvent event) {
        if (plot != null) {
            displayConfigurator("Plot frame configuration",
                    plot.getConfig(), DescriptorUtils.buildDescriptor(plot));
        }
    }

    @FXML
    private void onShowAll(ActionEvent event) {
        this.plot.getPlots().forEach(pl -> pl.configureValue("visible", true));
    }

    @FXML
    private void onHideAll(ActionEvent event) {
        this.plot.getPlots().forEach(pl -> pl.configureValue("visible", false));
    }

    /**
     * Set data loading progress. 1 means loading complete. Negative values correspond to indeterminate.
     * @param progress
     */
    public void setProgress(double progress){
        Platform.runLater(()->{
            this.progressProperty.set(progress);
        });
    }

    public DoubleProperty progressProperty() {
        return progressProperty;
    }

    protected class PlottableListCell extends ListCell<Plot> implements ConfigChangeListener {

        private HBox content;
        private CheckBox title;
        private Button configButton;

        /**
         * Configuration to which this cell is bound
         */
        private Configuration config;
        private Plottable plot;

        @Override
        protected synchronized void updateItem(Plot item, boolean empty) {
            super.updateItem(item, empty);

            //cleaning up after item with different config
            if (config != null) {
                config.removeObserver(this);
            }

            if (empty) {
                clearContent();
            } else {
                config = item.getConfig();
                config.addObserver(this, false);
                Platform.runLater(()->setContent(item));
            }
        }

        private void clearContent() {
            setText(null);
            setGraphic(null);
            plot = null;
        }

        private synchronized void setContent(Plottable item) {
            this.plot = item;
            setText(null);

            title = new CheckBox();
            title.setSelected(true);
            configButton = new Button("...");
            configButton.setMinWidth(0);
            Pane space = new Pane();
            HBox.setHgrow(space, Priority.ALWAYS);
            content = new HBox(title, space, configButton);
            HBox.setHgrow(content, Priority.ALWAYS);
            content.setMaxWidth(Double.MAX_VALUE);

//                title.textProperty().bindBidirectional(new ConfigStringValueProperty(config, "title"));
            title.setText(config.getString("title", item.getName()));
            title.setSelected(config.getBoolean("visible", true));
//                title.selectedProperty().bindBidirectional(new ConfigBooleanValueProperty(config, "visible"));
            title.selectedProperty().addListener((ObservableValue<? extends Boolean> observable, Boolean oldValue, Boolean newValue) -> {
                config.setValue("visible", newValue);
            });

            if (config.hasValue("color")) {
                title.setTextFill(Color.valueOf(config.getString("color")));
            }

            configButton.setOnAction((ActionEvent event) -> {
                displayConfigurator(item.getName() + " configuration", config, DescriptorUtils.buildDescriptor(item));
            });
            setGraphic(content);
        }

        @Override
        public synchronized void notifyValueChanged(String name, Value oldItem, Value newItem) {
            FXUtils.runNow(() -> {
                switch (name) {
                    case "title":
                        if(newItem == null){
                            title.setText(plot.getName());
                        } else {
                            title.setText(newItem.stringValue());
                        }
                        break;
                    case "color":
                        if(newItem == null){
                            title.setTextFill(Color.BLACK);
                        } else {
                            title.setTextFill(Color.valueOf(newItem.stringValue()));
                        }
                        break;
                    case "visible":
                        if(newItem == null){
                            title.setSelected(true);
                        } else {
                            title.setSelected(newItem.booleanValue());
                        }
                        break;
                    default:
                        break;
                }
            });
        }

        @Override
        public void notifyNodeChanged(String name, List<? extends Meta> oldItem, List<? extends Meta> newItem) {
            //ignore
        }

    }

}

File: kodex/kodex-fx/src/main/kotlin/hep/dataforge/kodex/fx/plots/PlotContainer.kt
package hep.dataforge.kodex.fx.plots

import hep.dataforge.description.DescriptorUtils
import hep.dataforge.description.NodeDescriptor
import hep.dataforge.fx.ApplicationSurrogate
import hep.dataforge.fx.FXObject
import hep.dataforge.fx.configuration.ConfigEditor
import hep.dataforge.kodex.fx.TableDisplay
import hep.dataforge.kodex.fx.dfIcon
import hep.dataforge.meta.Configuration
import hep.dataforge.meta.Meta
import hep.dataforge.names.Name
import hep.dataforge.plots.*
import hep.dataforge.plots.data.DataPlot
import javafx.application.Platform
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleDoubleProperty
import javafx.geometry.Orientation
import javafx.geometry.Pos
import javafx.scene.Node
import javafx.scene.Scene
import javafx.scene.control.TreeItem
import javafx.scene.image.ImageView
import javafx.scene.layout.Priority
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox
import javafx.scene.paint.Color
import javafx.scene.text.Font
import javafx.scene.text.TextAlignment
import javafx.stage.Stage
import tornadofx.*
import java.util.*


internal val defaultDisplay: (PlotFrame) -> Node = {
    if (it is FXObject) {
        it.fxNode
    } else {
        throw IllegalArgumentException("Can't display a plot frame since it is not an FX object")
    }
}

class PlotContainer(val frame: PlotFrame, display: (PlotFrame) -> Node = defaultDisplay) : Fragment(icon = ImageView(dfIcon)) {

    private val configWindows = HashMap<Configuration, Stage>()
    private val dataWindows = HashMap<Plot, Stage>()


    private val sideBarExpandedProperty = SimpleBooleanProperty(false)
    var sideBarExpanded by sideBarExpandedProperty

    private val sideBarPositionProperty = SimpleDoubleProperty(0.7)
    var sideBarPoistion by sideBarPositionProperty

    val progressProperty = SimpleDoubleProperty(1.0)
    var progress by progressProperty

    private lateinit var sidebar: VBox;

    override val root = borderpane {
        center {
            splitpane(orientation = Orientation.HORIZONTAL) {
                stackpane {
                    borderpane {
                        minHeight = 300.0
                        minWidth = 300.0
                        center = display(frame)
                    }
                    button {
                        graphicTextGap = 0.0
                        opacity = 0.4
                        textAlignment = TextAlignment.JUSTIFY
                        StackPane.setAlignment(this, Pos.TOP_RIGHT)
                        font = Font.font("System Bold", 12.0)
                        action {
                            sideBarExpanded = !sideBarExpanded;
                        }
                        sideBarExpandedProperty.addListener { _, _, expanded ->
                            if (expanded) {
                                setDividerPosition(0, sideBarPoistion);
                            } else {
                                setDividerPosition(0, 1.0);
                            }
                        }
                        textProperty().bind(
                                sideBarExpandedProperty.stringBinding {
                                    if (it == null || it) {
                                        ">>"
                                    } else {
                                        "<<"
                                    }
                                }
                        )
                    }

                    progressindicator(progressProperty) {
                        maxWidth = 50.0
                        prefWidth = 50.0
                        StackPane.setAlignment(this, Pos.CENTER)
                        visibleWhen(progressProperty.lessThan(1.0))
                    }

                }
                sidebar = vbox {
                    button(text = "Frame config") {
                        minWidth = 0.0
                        maxWidth = Double.MAX_VALUE
                        action {
                            displayConfigurator("Plot frame configuration", frame.config, DescriptorUtils.buildDescriptor(frame))
                        }
                    }
                    treeview<Plottable> {
                        minWidth = 0.0
                        root = PlotTreeItem(frame.plots);
                        vgrow = Priority.ALWAYS

                        //cell format
                        cellFormat { item ->
                            val cell = this
                            graphic = hbox {
                                hgrow = Priority.ALWAYS
                                checkbox(item.config.getString("title", item.name)) {
                                    minWidth = 0.0
                                    if (item == frame.plots) {
                                        text = "<<< All plots >>>"
                                    }
                                    isSelected = item.config.getBoolean("visible", true)
                                    selectedProperty().addListener { _, _, newValue ->
                                        item.config.setValue("visible", newValue)
                                    }


                                    if (frame is XYPlotFrame) {
                                        frame.getActualColor(getFullName(cell.treeItem)).ifPresent {
                                            textFill = Color.valueOf(it.stringValue())
                                        }
                                    } else if (item.config.hasValue("color")) {
                                        textFill = Color.valueOf(item.config.getString("color"))
                                    }

                                    item.config.addObserver { name, _, newItem ->
                                        when (name) {
                                            "title" -> text = if (newItem == null) {
                                                item.name
                                            } else {
                                                newItem.stringValue()
                                            }
                                            "color" -> textFill = if (newItem == null) {
                                                Color.BLACK
                                            } else {
                                                Color.valueOf(newItem.stringValue())
                                            }
                                            "visible" -> isSelected = newItem?.booleanValue() ?: true
                                        }
                                    }

                                    contextmenu {
                                        if (item is DataPlot) {
                                            item("Show data") {
                                                action {
                                                    displayData(item)
                                                }
                                            }
                                        } else if (item is PlotGroup) {
                                            item("Show all") {
                                                action { item.forEach { it.configureValue("visible", true) } }
                                            }
                                            item("Hide all") {
                                                action { item.forEach { it.configureValue("visible", false) } }
                                            }
                                        }
                                        if (cell.treeItem is PlotTreeItem) {
                                            item("Sort") {
                                                action { (cell.treeItem as PlotTreeItem).sort() }
                                            }
                                        }
                                    }
                                }

                                pane {
                                    hgrow = Priority.ALWAYS
                                }

                                button("...") {
                                    minWidth = 0.0
                                    action {
                                        displayConfigurator(item.name + " configuration", item.config, DescriptorUtils.buildDescriptor(item))
                                    }
                                }


                            }
                            text = null;
                        }

                    }
                }

                dividers[0].position = 1.0

                dividers[0].positionProperty().onChange {
                    if (it.toDouble() < 0.9) {
                        sideBarPositionProperty.set(it.toDouble())
                    }
                    sideBarExpanded = it.toDouble() < 0.99
                }

                this@borderpane.widthProperty().onChange {
                    if (sideBarExpanded) {
                        dividers[0].position = sideBarPoistion
                    } else {
                        dividers[0].position = 1.0
                    }
                }
            }
        }
    }

    fun addToSideBar(index: Int, vararg nodes: Node) {
        sidebar.children.addAll(index, Arrays.asList(*nodes))
    }

    fun addToSideBar(vararg nodes: Node) {
        sidebar.children.addAll(Arrays.asList(*nodes))
    }


    /**
     * Display configurator in separate scene
     *
     * @param config
     * @param desc
     */
    private fun displayConfigurator(header: String, config: Configuration, desc: NodeDescriptor) {
        configWindows.getOrPut(config) {
            Stage().apply {
                scene = Scene(ConfigEditor.build(config, desc))
                height = 400.0
                width = 400.0
                title = header
                setOnCloseRequest { configWindows.remove(config) }
                initOwner(root.scene.window)
            }
        }.apply {
            show()
            toFront()
        }
    }

    private fun displayData(plot: DataPlot) {
        dataWindows.getOrPut(plot) {
            Stage().apply {
                scene = Scene(TableDisplay(PlotUtils.extractData(plot, Meta.empty())).root)
                height = 400.0
                width = 400.0
                title = "Data: ${plot.name}"
                setOnCloseRequest { dataWindows.remove(plot) }
                initOwner(root.scene.window)
            }
        }.apply {
            show()
            toFront()
        }
    }

    private class PlotTreeItem(plot: Plottable, graphics: Node? = null) : TreeItem<Plottable>(plot, graphics) {
        init {
            (value as? PlotGroup)?.let { plt ->
                children.setAll(plt.children.map { PlotTreeItem(it) })
                plot.addListener { name ->
                    if (Name.of(name).length() == 1) {
                        (value as? PlotGroup)?.let { plt ->
                            children.setAll(plt.children.map { PlotTreeItem(it) })
                        }
                    }
                }
            }

            isExpanded = true
        }

        override fun isLeaf(): Boolean {
            return value !is PlotGroup
        }

        fun sort() {
            children.sortBy { it.value.name }
        }
    }

    private fun getFullName(item: TreeItem<Plottable>): String {
        if (item.parent == null || item.parent.value.name.isEmpty()) {
            return item.value.name;
        } else {
            return Name.joinString(getFullName(item.parent), item.value.name)
        }
    }


    companion object {

        /**
         * for testing
         */
        fun display(plot: PlotFrame, title: String = "", width: Double = 800.0, height: Double = 600.0) {
            Platform.setImplicitExit(false)
            ApplicationSurrogate.start()

            val container = PlotContainer(plot)

            Platform.runLater {
                ApplicationSurrogate.getStage().apply {
                    this.width = width
                    this.height = height
                    this.title = title
                    this.scene = Scene(container.root, width, height)
                    this.show()
                }
            }

            Platform.setImplicitExit(true)
        }
    }
}